crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+const BLOCK_SIZE: usize = METADATA_SIZE + DATA_SIZE;
+
+type TreeIndex = u32;
+type Parent = Option<TreeIndex>;
+type Hash = [u8; 32];
+type BlockBytes = [u8; BLOCK_SIZE];
+type KvId = u64;
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+#[repr(u8)]
+pub enum NodeType {
+    Internal = 0,
+    Leaf = 1,
+}
+
+impl NodeType {
+    pub fn from_u8(value: u8) -> Result<Self, String> {
+        // TODO: identify some useful structured serialization tooling we use
+        // TODO: find a better way to tie serialization values to enumerators
+        match value {
+            // ha!  feel free to laugh at this
+            x if (NodeType::Internal as u8 == x) => Ok(NodeType::Internal),
+            x if (NodeType::Leaf as u8 == x) => Ok(NodeType::Leaf),
+            other => panic!("unknown NodeType value: {other}"),
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            NodeType::Internal => NodeType::Internal as u8,
+            NodeType::Leaf => NodeType::Leaf as u8,
+        }
+    }
+}
+
+// impl NodeType {
+//     const TYPE_TO_VALUE: HashMap<NodeType, u8> = HashMap::from([
+//         (NodeType::Internal, 0),
+//         (NodeType::Leaf, 1),
+//     ]);
+//
+//     fn value(&self) -> u8 {
+//         let map = Self::TYPE_TO_VALUE;
+//         // TODO: this seems pretty clearly the wrong way, probably
+//         let value = map.get(self);
+//         if value.is_some() {
+//             return 3;
+//         }
+//         panic!("no value for NodeType: {self:?}");
+//     }
+// }
+
+fn internal_hash(left_hash: Hash, right_hash: Hash) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(b"\x02");
+    hasher.update(left_hash);
+    hasher.update(right_hash);
+
+    hasher.finalize()
+}
+
+pub struct DotLines {
+    nodes: Vec<String>,
+    connections: Vec<String>,
+    pair_boxes: Vec<String>,
+}
+
+impl Default for DotLines {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DotLines {
+    pub fn new() -> Self {
+        Self {
+            nodes: vec![],
+            connections: vec![],
+            pair_boxes: vec![],
+        }
+    }
+
+    pub fn push(&mut self, mut other: DotLines) {
+        self.nodes.append(&mut other.nodes);
+        self.connections.append(&mut other.connections);
+        self.pair_boxes.append(&mut other.pair_boxes);
+    }
+
+    pub fn dump(&mut self) -> String {
+        // TODO: consuming itself, secretly
+        let mut result = vec!["digraph {".to_string()];
+        result.append(&mut self.nodes);
+        result.append(&mut self.connections);
+        result.append(&mut self.pair_boxes);
+        result.push("}".to_string());
+
+        result.join("\n")
+    }
+}
+
+const NULL_PARENT: TreeIndex = 0xffff_ffffu32;
+
+#[derive(Debug, PartialEq)]
+pub struct NodeMetadata {
+    pub node_type: NodeType,
+    pub dirty: bool,
+}
+
+impl NodeMetadata {
+    pub fn from_bytes(blob: [u8; METADATA_SIZE]) -> Result<Self, String> {
```

Comment:
that was a thought and i haven't looked at how that would work out.  i stuck with this low level for at least awhile because i wanted to get some familiarity at this layer.  i (think i) do in places actually deserialize just individual fields which i'm guessing streamable doesn't support.  so it may be that i have a (pre-optimization) reason to not use streamable.  but yes, this is a good thing to reconsider prior to considering this code 'done'.